pkg/sign/content.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2024 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sign
+
+import (
+	"fmt"
+
+	protobundle "github.com/sigstore/protobuf-specs/gen/pb-go/bundle/v1"
+	protocommon "github.com/sigstore/protobuf-specs/gen/pb-go/common/v1"
+	protodsse "github.com/sigstore/protobuf-specs/gen/pb-go/dsse"
+)
+
+type Content interface {
+	// Return the data to be signed
+	PreAuthEncoding() []byte
+	// Returns something that satisfies protobundle.isBundle_Content
+	Bundle(bundle *protobundle.Bundle, hashAlgorithm protocommon.HashAlgorithm, digest []byte, signature []byte)
```

Comment:
Yeah, I don't love having `hashAlgorithm` and `digest` be arguments to `Bundle()` either.

The problem I'm trying to solve is that the `Signer` isn't necessarily going to know what kind of content it's signing, and the `hashAlgorithm` (and associated `digest`) comes from the `Keypair`, not the `Content`.

Actually, now that I type this out, maybe instead of having a public `PreAuthEncoding()` method, `Content` should have a `SetHashAlgorithm` that stores the digest and signature on itself, ready for a call to `Bundle()`.

Reply:
Okay I took a pass at having a cleaner separation of responsibilities between `Content` and `Keypair`. The dataflow of the desired hash algorithm still feels a little weird, but the function arguments are more straightforward.